### PR TITLE
Better webgl support

### PR DESCRIPTION
**Pull request by omni-viral**
Fixes #issue
PR checklist:
- [x] `make` succeeds
- [ ] `make reftests` succeeds
- [ ] tested examples with the following backends:
- [ ] `rustfmt` run on changed code

### CODE REVIEW

File: src/backend/gl/src/command.rs [hunk 1]
```diff
@@ -420,6 +424,10 @@ impl CommandBuffer {
 
             match vertex_buffer_descs.get(binding) {
                 Some(&Some(desc)) => {
+                    if let pso::VertexInputRate::Instance(count) = desc.rate {
+                        attribute.offset += desc.stride * first_instance / count as u32;
```
Comment (msiglreith) [inline src/backend/gl/src/command.rs]: afaik first instance is independent of the instance divisor

Comment (omni-viral) [reply src/backend/gl/src/command.rs]: There's no such thing in Vulkan.
So, what behavior gfx should have?

File: src/backend/gl/src/device.rs [hunk 2]
```diff
@@ -823,9 +831,17 @@ impl d::Device<B> for Device {
...
-                    let loc = gl.get_uniform_location(name, bname);
-                    gl.uniform_1_i32(loc, *binding as _);
+                for (bname, (btype, binding)) in name_binding_map.iter() {
+                    match btype {
+                        n::BindingTypes::Images => {
+                            let loc = gl.get_uniform_location(name, bname);
+                            gl.uniform_1_i32(loc, *binding as _);
+                        }
+                        n::BindingTypes::UniformBuffers | n::BindingTypes::StorageBuffers => {
+                            let index = gl.get_uniform_block_index(name, bname).unwrap();
```
Comment (kvark) [inline src/backend/gl/src/device.rs]: I don't think `StorageBuffer` would end up as a uniform block in GLSL. We should limit this match arm to `UniformBuffers` only

File: src/backend/gl/src/command.rs (hunk 1 above)
Comment (msiglreith) [reply src/backend/gl/src/command.rs]: It should behave the same way as defined the vk divisor extension:
```c
if (divisor == 0)
    vertexOffset = firstInstance * bindingDesc.stride;
else
    vertexOffset = (firstInstance + ((instanceIndex - firstInstance) / divisor)) *
bindingDesc.stride;
```

File: src/backend/gl/src/device.rs (hunk 2 above)
Comment (omni-viral) [reply src/backend/gl/src/device.rs]: Yeah, I see.
But is it possible to retrieve storage block index by name?

**Comment by kvark** [comment]
I assume the winit git dependency is temporary?
bors r+

**Comment by kvark** [comment]
bors retry

**Comment by omni-viral** [comment]
@kvark yes, winit git depenency is temporary. Probably until next release.

**Comment by kvark** [comment]
some sort of a network outage on CI :(

**Comment by kvark** [comment]
Bors retry

**Comment by grovesNL** [comment]
Looks like the CI failure is legitimate and caused by the winit branch:
```
error[E0432]: unresolved import `web_sys`

  --> C:\Users\travis\.cargo\git\checkouts\winit-c2fdb27092aba5a7\dbdde3d\src\platform\web.rs:10:5

   |

10 | use web_sys::HtmlCanvasElement;

   |     ^^^^^^^ use of undeclared type or module `web_sys`
```
cc @omni-viral

**Comment by omni-viral** [comment]
Oh, right, my winit fix wasn't merged yet.

**Comment by grovesNL** [comment]
@omni-viral looks like the winit branch has been merged, but this branch needs to be rebased

**Comment by kvark** [comment]
Please rebase

**Comment by kvark** [comment]
@omni-viral can we revive this?

**Comment by omni-viral** [comment]
@kvark working on it
It seems that winit web support will take another month or so, IDK.
So I will stick to custom codepath as it is right now in master.
Expect update on weekend.

**Pull request merged**